test/k8sT/Services.go

```diff
@@ -721,6 +723,107 @@ var _ = Describe("K8sServicesTest", func() {
+					dstPod = pod
+				} else {
+					// Check that destination pod is always the same
+					Expect(dstPod).To(Equal(pod))
+				}
+			}
+
+			// Delete the pod, and check that a new backend is chosen
+			kubectl.DeleteResource("pod", dstPod).ExpectSuccess("Unable to delete %s pod", dstPod)
+			kubectl.WaitPodDeleted(dstPod)
+			// Unfortunately, it takes a while until cilium-agent receives Endpoint
+			// update event which triggers a removal of the deleted dstPod from
+			// the affinity and the service BPF maps. Therefore, the bellow requests
+			// are flaky.
+			// TODO(brb) don't sleep, instead wait for Endpoint obj update (might be complicated though)
+			time.Sleep(7 * time.Second)
+
+			for i := 1; i <= count; i++ {
+				res, err := kubectl.ExecInHostNetNS(context.TODO(), externalNodeName, cmd)
+				Expect(err).Should(BeNil(), "Cannot exec in %s host netns", externalNodeName)
+				ExpectWithOffset(1, res).Should(helpers.CMDSuccess(),
+					"Cannot connect to service %q from outside cluster (%d/%d) after restart", httpURL, i, count)
+				pod := strings.TrimSpace(strings.Split(res.GetStdOut(), ": ")[1])
+				if i == 1 {
+					// Retrieve the destination pod from the first request
+					Expect(dstPod).ShouldNot(Equal(pod))
+					dstPod = pod
+				} else {
+					// Check that destination pod is always the same
+					Expect(dstPod).To(Equal(pod))
+				}
+			}
+
+			// Now check from inside the cluster (bpf_sock.c)
+			By("Making %d HTTP requests from %q to %q", count, testDSClient, httpURL)
+
+			pods, err := kubectl.GetPodNames(helpers.DefaultNamespace, testDSClient)
+			ExpectWithOffset(1, err).Should(BeNil(), "cannot retrieve pod names by filter %q", testDSClient)
+
+			for i := 1; i <= count; i++ {
+				res := kubectl.ExecPodCmd(helpers.DefaultNamespace, pods[1], cmd)
+				ExpectWithOffset(1, res).Should(helpers.CMDSuccess(),
+					"Pod %q can not connect to service %q (%d/%d)", pods[1], httpURL, i, count)
+				pod := strings.TrimSpace(strings.Split(res.GetStdOut(), ": ")[1])
+				if i == 1 {
+					// Retrieve the destination pod from the first request
+					dstPod = pod
+				} else {
+					// Check that destination pod is always the same
+					Expect(dstPod).To(Equal(pod))
+				}
+			}
+
+			// Again, delete the backend and check that new is chosen
+			kubectl.DeleteResource("pod", dstPod).ExpectSuccess("Unable to delete %s pod", dstPod)
+			kubectl.WaitPodDeleted(dstPod)
+			time.Sleep(7 * time.Second)
+
+			for i := 1; i <= count; i++ {
+				res := kubectl.ExecPodCmd(helpers.DefaultNamespace, pods[1], cmd)
+				ExpectWithOffset(1, res).Should(helpers.CMDSuccess(),
+					"Pod %q can not connect to service %q (%d/%d)", pods[1], httpURL, i, count)
+				pod := strings.TrimSpace(strings.Split(res.GetStdOut(), ": ")[1])
+				if i == 1 {
+					// Retrieve the destination pod from the first request
+					dstPod = pod
+				} else {
+					// Check that destination pod is always the same
+					Expect(dstPod).To(Equal(pod))
+				}
```

Comment:
Suggestion: some of this code is duplicated several times above. Maybe a candidate to put into a helper function? For me, I like to write test code which has the action and then the assertion explicitly following each other, but we can probably find a good function name that can signal what's happening here.

Reply:
In the beginning, I extended `doRequestsFromThirdHostWithLocalPort()` to do the check, but that just hindered readability of the test helper. Anyway, I will move the check to a separate helper function and based on a param, requests will be sent either from a third host or a pod.